File: 0149_maxPointsOnALine/main.cpp
#include <vector>
#include <unordered_map>
#include <climits>
#include <algorithm>
class Solution {
public:
    int maxPoints(std::vector<std::vector<int>>& points) {
        if(points.size() <= 2)
            return points.size();

        int res = 0;
        for(int i = 0; i < points.size(); ++i) {
            std::unordered_map<double, int> countMap;
            for(int j = i + 1; j < points.size(); ++j) {
                double dy = points[i][1] - points[j][1];
                double dx = points[i][0] - points[j][0];
                double slope;
                if(dx == 0)
                    slope = std::numeric_limits<double>::max();
                else
                    slope = dy / dx;
                ++countMap[slope];
                res = std::max(res, countMap[slope] + 1);
            }
        }
        return res;
    }
};

File: 0149_maxPointsOnALine/main.py
from typing import List
import collections


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        if len(points) <= 2:
            return len(points)
        
        res = 0
        for i in range(len(points)):
            countMap = collections.defaultdict(int,)
            for j in range(i + 1, len(points)):
                y = points[i][1] - points[j][1]
                x = points[i][0] - points[j][0]
                slope = y / x if x != 0 else float('inf')

                countMap[slope] += 1
                res = max(res, countMap[slope] + 1)
        return res                
